processing/src/main/java/org/apache/druid/math/expr/Function.java

```diff
@@ -3803,16 +3854,56 @@ public ExpressionType getOutputType(Expr.InputBindingInspector inspector, List<E
     }
 
     @Override
-    ExprEval doApply(ExprEval lhsExpr, ExprEval rhsExpr)
+    public ExprEval apply(List<Expr> args, Expr.ObjectBinding bindings)
     {
-      final Object[] array1 = lhsExpr.asArray();
-      final List<Object> array2 = Arrays.asList(rhsExpr.asArray());
+      final ExprEval arrayExpr1 = args.get(0).eval(bindings);
+      final ExprEval arrayExpr2 = args.get(1).eval(bindings);
+
+      final Object[] array1 = arrayExpr1.asArray();
+      final Object[] array2 = arrayExpr2.asArray();
+      if (array1 == null) {
+        return ExprEval.ofLong(null);
```

Comment:
This makes me wonder what happens when the rhs array includes null.

Consider `MV_CONTAINS(x, ARRAY[NULL, 'abc', 'def'])`, when `x` is a MVD where one row contains a single `NULL`. Is that equivalent to `x IS NULL OR MV_CONTAINS(x, ARRAY['abc', 'def'])` (which returns `TRUE`) or is it equivalent to `MV_CONTAINS(x, ARRAY['abc', 'def'])`?

For that matter, consider `MV_CONTAINS(x, ARRAY['abc', 'def'])`. Does it return `FALSE` as if `x` was treated like `ARRAY[NULL]` (a nonnull array containing null)? Or does it return `UNKNOWN` as if `x` was `NULL` itself?

Do the answers to these questions depend on how the `MV_CONTAINS` is planned— whether it ends up as an `array_contains` or a filter or something else?

Reply:
heh, good question. So if I remember correctly, by default for an expression like `MV_CONTAINS` the value `NULL` would just be `NULL`, so this would be checking `MV_CONTAINS(NULL, ARRAY[NULL, 'abc', 'def'])`.

This is controlled https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/segment/virtual/ExpressionSelectors.java#L344 where the 'homogenize' flag which translates stuff like `NULL` and `[]` into `[NULL]` is only done for scalar expressions which are transformed to apply the expression across the MVD, but since `MV_CONTAINS` is `array_contains` native expression, this transformation does not happen because we are using the MVD as an array instead.

So, in all of these cases it returns `UNKNOWN` when evaluated as an expression.

Looking at MV_CONTAINS as a filter in the sql operator conversions, it should behave the same way except in one case I think due to how we rewrite it into native filters. The case where I think it would be different is `MV_CONTAINS(mvd, [NULL])` would rewrite into basically `mvd is null`, which would return true instead of unknown. Maybe we should prevent the `MV_CONTAINS(mvd, [NULL])` case from being allowed to rewrite in this case?

Reply:
oh to clarify a bit more, `MV_CONTAINS(x, ARRAY[NULL, 'abc', 'def'])` gets rewritten at sql planner into `x is null AND x = 'abc' AND x = 'def'` if `x` is a string (contains is an AND operation, not an OR operation, overlap is the OR operation though we use an IN filter instead). `array_contains`/`mv_contains` are allowed to check if an array contains a null element.